sshd-common/src/main/java/org/apache/sshd/common/cipher/BaseGCMCipher.java

```diff
@@ -78,19 +78,25 @@ public void update(byte[] input, int inputOffset, int inputLen) throws Exception
      */
     protected static class CounterGCMParameterSpec extends GCMParameterSpec {
         protected final byte[] iv;
+        protected final long initialCounter;
 
         protected CounterGCMParameterSpec(int tLen, byte[] src) {
             super(tLen, src);
             if (src.length != 12) {
                 throw new IllegalArgumentException("GCM nonce must be 12 bytes, but given len=" + src.length);
             }
             iv = src.clone();
+            initialCounter = BufferUtils.getLong(iv, iv.length - Long.BYTES, Long.BYTES);
         }
 
         protected void incrementCounter() {
             int off = iv.length - Long.BYTES;
             long counter = BufferUtils.getLong(iv, off, Long.BYTES);
-            BufferUtils.putLong(Math.addExact(counter, 1L), iv, off, Long.BYTES);
+            long newCounter = counter + 1L;
+            if (newCounter == initialCounter) {
```

Comment:
There doesn't seem to be any harm in adding this check, but the way that SSH works would force a rekey event well before this counter would ever loop back to where it started (rekey events happen far more often than every 2^64 packets).

Reply:
Yes, I agree, but I figured it would be better to be safer than sorry incase there was some sort of misuse or issue in the future, since IV reuse could be disastrous.